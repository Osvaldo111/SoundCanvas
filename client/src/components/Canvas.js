import React from "react";
import "../style/canvas.css";
import { connect } from "react-redux";
import { changeCanvasWidth } from "../actions";
import { isMobileBttnPress } from "../actions";
import { displayMobileBttn } from "../actions";
import recordBttn from "../images/recordBttn.svg";
// import audioUtilities from "../utilities/resizeCanvas";
import audioUtilities from "../utilities/recording";
import graphs from "../utilities/drawGraphClassic";

import mobilBttn from "../images/menu-bttn.svg";
/**
 * author: Osvaldo Carrillo
 * Date: 02/08/2020
 * This class represents the canvas on which the
 * sound wave is drawn.
 */
class Canvas extends React.Component {
  constructor(props) {
    super(props);
    this.canvasRef = React.createRef();
    this.state = {
      displayBttn: true,
      canvasWidth: 600,
      canvasHeight: 600,
      recorderBttn: false,
      arrayOfAmplitud: [],
      developmentVar: "example"
    };
  }

  componentDidMount() {
    this.processAudio(); // Development: Change on click
    this.drawClassicGraph();
  }

  drawClassicGraph = () => {
    const canvas = this.canvasRef.current;
    var arrAmplitud = [10, 20, 30, 100];
    graphs.drawGraphInCanvas(canvas, this.state.arrayOfAmplitud);
    requestAnimationFrame(this.drawClassicGraph);
  };

  componentDidUpdate(prevProps, prevState) {
    const currentMobileNavBttn = this.props.canvasProps.displayMobileBttn;
    const prevMobileNavBttn = prevProps.canvasProps.displayMobileBttn;

    if (currentMobileNavBttn !== prevMobileNavBttn) {
      if (currentMobileNavBttn) {
        this.setState({ displayBttn: true });
        //Reset reducer init
        this.props.displayMobileBttn(false);
      }
    }
  }

  /**
   * This function is designed to check and as for permission
   * to record the audio from the user. This implemets the
   * getUserMedia method from the Media devices to received the
   * MediaStream object which consists of several audio tracks.
   * Also, this function connects the stream to the AudioContext
   * and the AnalyserNode to get the frequency data. This function sets the
   * Audio Context interface that represents an audio-proccesing graph build
   * from audio modules.
   * This interface uses the createMediaStreamSource method to create a
   * MediaElementAudioSourceNode from which the audio can be manipulated.
   * The AnalyserNode interface represents a node that provides real-time
   * frequency and time-domain analysis information.
   * https://developer.mozilla.org/en-US/docs/Web/API/MediaDevices
   * https://developer.mozilla.org/en-US/docs/Web/API/MediaDevices/getUserMedia
   * https://developer.mozilla.org/en-US/docs/Web/API/MediaStream
   */
  processAudio = () => {
    if (navigator.mediaDevices && navigator.mediaDevices.getUserMedia) {
      console.log("getUserMedia supported.");
      navigator.mediaDevices
        .getUserMedia(
          // constraints - only audio needed for this app
          {
            audio: true
          }
        )
        // Success callback
        .then(stream => {
          // Setting the Audio Context parameters
          console.log("Context");
          var audioCtx = new (window.AudioContext ||
            window.webkitAudioContext)();
          var analyser = audioCtx.createAnalyser();
          analyser.minDecibels = -90;
          analyser.maxDecibels = -10;
          analyser.smoothingTimeConstant = 0;
          var source = audioCtx.createMediaStreamSource(stream);
          source.connect(analyser);
          // The fftSize is converted half the value
          // for technical resons.
          analyser.fftSize = 32;
          var bufferLength = analyser.frequencyBinCount;
          var dataArray = new Uint8Array(bufferLength);

          // Connection to Visualize the Data.
          this.connectionToStreamSource(analyser, dataArray, audioCtx);

          // drawGraphInCanvas();
        })
        // Error callback
        .catch(function(err) {
          console.log("The following getUserMedia error occured: " + err);
        });
    } else {
      console.log("getUserMedia not supported on your browser!");
    }
  };

  connectionToStreamSource = (analyser, dataArray, audioCtx) => {
    var counterOfInterval = 0;
    var myInterval = setInterval(() => {
      counterOfInterval++;
      var frequencyData = audioUtilities.getFrequencyData(analyser, dataArray);
      this.setState({ arrayOfAmplitud: frequencyData }, () => {
        console.log("After setting the Array", frequencyData);
      });
      if (counterOfInterval === 5) {
        audioCtx.close().then(() => {
          console.log("Stop Recording");
        });
        clearInterval(myInterval);
      }
    }, 1000);
  };

  pressBttn = () => {
    alert();
  };

  isMobileBttnPress = () => {
    //Reducer bttn is press.
    this.props.isMobileBttnPress(true);
    //Reducer display bttn
    // this.props.displayMobileBttn(false);
    this.setState({ displayBttn: false });
  };

  isRecorderBttnPressed = () => {
    const { recorderBttn } = this.state;
    if (recorderBttn === false) {
      this.setState({ recorderBttn: true }, () => {
        // recordAudio();
      });
    }
  };

  // componentDidMount() {
  // const canvas = resizeCanvas(this.canvasRef);
  // // console.log(canvas.width);
  // // console.log(canvas.height);
  // if (canvas.width <= 425) {
  //   const newCanvasWidth = canvas.width - 75;
  //   // console.log(newCanvasWidth);
  //   this.setState({
  //     canvasWidth: newCanvasWidth,
  //     canvasHeight: newCanvasWidth
  //   });
  // }
  // window.addEventListener("resize", () => {
  //   const canvas = resizeCanvas(this.canvasRef);
  //   // console.log(canvas.width);
  //   // console.log(canvas.height);
  //   if (canvas.width <= 425) {
  //     const newCanvasWidth = canvas.width - 75;
  //     console.log(newCanvasWidth);
  //     this.setState({
  //       canvasWidth: newCanvasWidth,
  //       canvasHeight: newCanvasWidth
  //     });
  //   }
  // });
  // }

  render() {
    const { canvasWidth, canvasHeight } = this.state;
    const { displayBttn, arrayOfAmplitud } = this.state;
    return (
      <div className="canvasContainer">
        <canvas
          className="canvas"
          width={canvasWidth}
          height={canvasHeight}
          /* style={{ width: canvasWidth, height: canvasHeight }} */
          ref={this.canvasRef}
        ></canvas>
        <div className="recordBttn">
          <img src={recordBttn} alt="" onClick={this.isRecorderBttnPressed} />
        </div>
        <div className="resetBttnMobile">
          <button onClick={this.pressBttn}>Reset</button>
        </div>
        <div className="mobilBttn">
          <img
            onClick={this.isMobileBttnPress}
            src={mobilBttn}
            alt=""
            style={{ display: displayBttn ? "flex" : "none" }}
          />
        </div>
        <p>{arrayOfAmplitud.length}</p>
      </div>
    );
  }
}

function mapStateToProps(state) {
  return { canvasProps: state.canvas };
}

const mapDispatchToProps = {
  changeCanvasWidth,
  isMobileBttnPress,
  displayMobileBttn
};

export default connect(mapStateToProps, mapDispatchToProps)(Canvas);
