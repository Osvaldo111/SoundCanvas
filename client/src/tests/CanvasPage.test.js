import React from "react";
import ReactDOM from "react-dom";
import { configure } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import Canvas from "../components/Canvas";
import { mount, shallow } from "enzyme";
//Redux
import { Provider } from "react-redux";
import store from "../reducers/store";
import { createSerializer } from "enzyme-to-json";
import { spy } from "sinon";
configure({ adapter: new Adapter() });

expect.addSnapshotSerializer(createSerializer({ mode: "deep" }));

describe("Canvas page <Canvas />", () => {
  it("renders correctly", () => {
    const wrapper = mount(
      <Provider store={store}>
        <Canvas />
      </Provider>
    );
    expect(wrapper).toMatchSnapshot();
  });
});
