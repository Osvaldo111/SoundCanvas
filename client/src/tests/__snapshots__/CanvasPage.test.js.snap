// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Canvas page <Canvas /> renders correctly 1`] = `
<Connect(Canvas)>
  <div
    className="canvasContainer"
  >
    <div
      className="canvasSection"
    >
      <canvas
        className="canvas"
        height={300}
        style={
          Object {
            "backgroundColor": "#fff",
          }
        }
        width={300}
      />
      <input
        className="inputTextCanvas"
        onChange={[Function]}
        onKeyDown={[Function]}
        onMouseDown={[Function]}
        onTouchStart={[Function]}
        size="8"
        style={
          Object {
            "display": "none",
            "fontSize": 14,
          }
        }
        type="text"
        value="TEXT HERE"
      />
    </div>
    <div
      className="recordBttn"
    >
      <img
        alt=""
        onClick={[Function]}
        src="recordBttn.svg"
        style={
          Object {
            "opacity": "",
          }
        }
      />
    </div>
    <div
      className="canvBttnOptions"
    >
      <div
        className="resetBttnMobile"
      >
        <button
          onClick={[Function]}
          style={
            Object {
              "opacity": "0.6",
            }
          }
        >
          Reset
        </button>
      </div>
      <a
        className="mobileBttnEffect"
        download="myImage.jpg"
        id="download"
        onClick={[Function]}
      >
        <div
          id="spinAnimMob"
        />
        <p>
          send
        </p>
      </a>
      <a
        className="mobileBttnEffect"
        download="myImage.jpg"
        id="download"
        onClick={[Function]}
      >
        <div
          id="spinAnimMob"
        />
        <p>
          Download
        </p>
      </a>
    </div>
    <div
      className="mobilBttn"
    >
      <img
        alt=""
        onClick={[Function]}
        src="menu-bttn.svg"
        style={
          Object {
            "display": "flex",
          }
        }
      />
    </div>
    <div
      style={
        Object {
          "display": "none",
        }
      }
    >
      <a
        download="myImage.jpg"
        href=""
      />
    </div>
    <Connect(SendForm)
      disableBttn={false}
      display={false}
      handleDisplay={[Function]}
      handleSend={[Function]}
      message=""
    >
      <form
        className="overSendCanv"
        onSubmit={[Function]}
        style={
          Object {
            "opacity": "",
            "visibility": "",
          }
        }
      >
        <div
          className="sendCardConta"
        >
          <div
            className="sendCardForm"
          >
            <div
              className="sendCarFormTitle"
            >
              Send Canvas
            </div>
            <div
              className="sendCanvFormBdy"
            >
              <div>
                <input
                  onChange={[Function]}
                  placeholder="Email Address"
                  type="email"
                  value=""
                />
              </div>
            </div>
            <div
              className="ruleLnFrm"
            />
            <div
              className="sendCvaFrmFttr"
            >
              <input
                disabled={true}
                style={
                  Object {
                    "cursor": "default",
                    "opacity": "0.5",
                  }
                }
                type="submit"
                value="Send"
              />
              <a
                onClick={[Function]}
              >
                Close
              </a>
              <div />
            </div>
          </div>
        </div>
      </form>
    </Connect(SendForm)>
  </div>
</Connect(Canvas)>
`;
